pkg/ruler/merger.go

```diff
@@ -0,0 +1,34 @@
+package ruler
+
+import (
+	"time"
+
+	promRules "github.com/prometheus/prometheus/rules"
+)
+
+// mergeGroupStateDesc removes duplicates from the provided []*GroupStateDesc by keeping the GroupStateDesc with the
+// latest information. It uses the EvaluationTimestamp of the GroupStateDesc and the EvaluationTimestamp of the
+// ActiveRules in a GroupStateDesc to determine the which GroupStateDesc has the latest information.
+func mergeGroupStateDesc(in []*GroupStateDesc) []*GroupStateDesc {
+	states := make(map[string]*GroupStateDesc)
+	rgTime := make(map[string]time.Time)
+	for _, state := range in {
+		latestTs := state.EvaluationTimestamp
+		for _, r := range state.ActiveRules {
+			if latestTs.Before(r.EvaluationTimestamp) {
+				latestTs = r.EvaluationTimestamp
+			}
+		}
```

Comment:
with this PR we are still evaluating all the rules in the replicas as well? if we are not evaluating, this is not required for the instances that are not primary right?

if yes, should we only consider the timestamp of the group? reason why is because the timestamp in the rule is from the time the rule is being evaluated, which can be delayed -> you might be discarding a GroupStateDesc from a good instance in favor of delayed GroupStateDesc just because a rule took a long time to evaluate?

Reply:
>with this PR we are still evaluating all the rules in the replicas as well? if we are not evaluating, this is not required for the instances that are not primary right?

only the primary instance will evaluate the rule group. so only the data from the primary instance will have the latest EvaluationTimestamps. 

The `EvaluationTimestamp` of `RuleGroupDesc` is based on the [GetLastEvaluation](https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler.go#L809) of the [Group](https://github.com/prometheus/prometheus/blob/main/rules/group.go#L322C17-L326). The [setter](https://github.com/prometheus/prometheus/blob/main/rules/group.go#L329-L335) of this value is only called after the [Group finished evaluation](https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L86). This if you have a rule group that takes 5 minutes to evaluate maybe because the rules inside it are slow or there are just way too many rules, it would take 5 minutes since the first run of the Group before `EvaluationTimestamp` is populated. If we look only at the Groups `EvaluationTimestamp` we could be returning the data from the backup (because they will have the same value) during this 5 minutes which will not be accurate because within this 5 minutes it could be that some rules within the Group has already evaluated and has a state set. This is why I opted to checking the Rule's `EvaluationTimestamp` as well because that timestamp gets updated [per rule evaluation](https://github.com/prometheus/prometheus/blob/main/rules/group.go#L457-L464) and does not wait for the entire Group to finish.